Blocks: Render columns as nested column (wrapper)

aduth opened the pull request:
Closes #5351

This pull request seeks to address an issue where the current Columns block implementation using grid is insufficient for supporting columns of varying height where it is expected for blocks within a column to stack atop each other, thus resulting in undesirable empty space.

Before|After
---|---
![Before](https://user-images.githubusercontent.com/1779930/41172752-d45101b6-6b22-11e8-9619-4ade4c3d4315.png)|![After](https://user-images.githubusercontent.com/1779930/41172652-83195ff0-6b22-11e8-98eb-c417771c4c14.png)

**Implementation notes:**

This explores one option of addressing #5351 which was considered at https://github.com/WordPress/gutenberg/issues/5351#issuecomment-375795651, where in order to introduce a wrapper element for each column, the columns themselves are an intermediary block. This is possible through new APIs for `supportedBlocks` of a block type (#6753) and `allowedBlocks` and `template` of an `InnerBlocks` (#6546). Thus, a "Column" block is only valid within a Columns wrapper, and conversely a Column block is the only valid child of a Columns block. This will require that the Columns block fully manages its own template, which is only partially implemented thus far.

This may allow for the eventual removal of `layouts` concept altogether.

The primary alternative here is to keep the distinction of "grouped" and "ungrouped" layouts ([documentation](https://github.com/WordPress/gutenberg/tree/master/editor/components/inner-blocks)), and automatically insert a wrapper element for grouped layouts. This is challenging both in the further inconsistency between the two types of layouts, and in the current unavailability of `layouts` anywhere outside of the `edit` implementation where it is passed (it would likely need to become a top-level property of the block).

**Testing instructions:**

Verify that you can create columns, insert blocks within (via slash insertion), and that within columns, the individual blocks stack neatly atop each other when previewed on the front of your site.

aduth committed:
Blocks: Render columns as nested column (wrapper)

chrisvanpatten commented:
The UX with the nested blocks is a little funky, and changing the number of columns doesn't work yet, but it does feel like a step in the right direction.

Excited to see this evolve further!

SuperGeniusZeb commented:
I made an issue suggesting this sort of thing a while back: #6461. I backed off the idea after looking into the [Layout block mockup](https://github.com/WordPress/gutenberg/issues/5351#issuecomment-371885817) in #5351, which I thought would be a better solution that could replace the Columns block entirely. I am a bit surprised the each-column-is-a-block direction is actually being pursued.

Some questions I have about this approach:
- Is it possible to implement unequal-width columns? This is essential for making the Columns block useful.
- Is it possible to implement responsive columns? This is even more essential.
- Is it possible to have a horizontal sibling inserter to insert additional Column blocks? Currently, the UI/UX for inserting a Column into Columns is pretty confusing, because you insert it via the sibling inserter above one of the existing columns, but when you do so, the Column is inserted horizontally. I imagine a sibling inserter rotated 90 degrees and usable between two Column blocks would be more ideal.

Notably, the <samp>Columns</samp> slider in the Columns block does nothing now. If this implementation is used, would this slider be removed and replaced with something like a horizontal sibling inserter?

Also, due to the current way the sibling inserter works, you can insert Paragraph blocks directly into the Columns block, which produces very interesting results. Obviously, this is not intentional, but I thought I would point it out because of the interesting way the widths of the blocks are determined based on the content.
![image](https://user-images.githubusercontent.com/19592990/41177331-3e52398a-6b29-11e8-9346-5f6e5e7ef868.png)

chrisvanpatten commented:
The more I use it, the more concerned I am about the UX. I think the conceptual idea of a "Columns" block with "Column" children makes a sort of academic sense, but I think just introduces more confusion.

Do new columns get added via the inserter somehow? How does that inserter look, especially in more complex column/row layouts? Or if not, and the Columns block manages its own children, how is that articulated to users (there are nested blocks, but the blocks aren't really like other blocks, because you don't manage them the way other blocks are managed, etc.)? How do you coordinate the relationship from block to block? Would there need to be new APIs to help blocks manage their children and keep them in sync?

I definitely can be convinced there's a way to make this work, and I can appreciate the sort of "semantic purity" that a "Columns > Column > Your Blocks Here" solution brings, but I think it introduces more points of confusion from an end-user perspective.

> The primary alternative here is to keep the distinction of "grouped" and "ungrouped" layouts (documentation), and automatically insert a wrapper element for grouped layouts. This is challenging both in the further inconsistency between the two types of layouts, and in the current unavailability of layouts anywhere outside of the edit implementation where it is passed (it would likely need to become a top-level property of the block).

Inconsistency between the two types of layouts (e.g. actually grouping grouped layouts), and layouts as a top-level block property, both seem like good ideas :)

GlennMartin1 commented:
Can this make it into 3.1?

eddr commented:
Also, please consider allowing for a flexbox layout instead of grid (as a selectable option)

aduth commented:
The changes here include revisions for the columns block to use Flex instead of Grid.

https://github.com/WordPress/gutenberg/blob/a8d9202af03ded828814c7b7ce4805c96d55e909/core-blocks/columns/style.scss#L1-L7

paulwilde commented:
This would fix #5935.

My only concern is that it's very time-consuming to build a layout if you are required to individually add a column block for each column before you can add other columns. If this is handled automatically then disregard this comment as I haven't tested this PR out.

aduth committed:
Block API: Synchronize blocks as unmodified if no template provided

aduth committed:
Inner Blocks: Keep inner blocks in sync with template

aduth commented:
Rebased and pushed a few changes:

- The Column block was previously not restricted to its Columns parent, and therefore would be allowed to be inserted standalone outside this context. This has been resolved in b725803.
- `InnerBlocks` now continuously keeps its template in sync. When the user uses the Columns range slider to change the number of columns, the block view updates accordingly.

There's still a couple awkwardnesses:

- If you select the Column block (by pressing ArrowUp from a paragraph within it), then use the inserter, you have the option to insert a Column block. This doesn't do anything, and is not expected to, since column count is managed by the inspector control, but it may be unexpected for a user to encounter.

GlennMartin1 commented:
Any chance this could be a part of 3.1?

robbisy commented:
Is there any plan to add a wrapper option for innerblock that could wrap each of this columns in a component. I've seen there's a plan to add wrapping grouped layout in innerblocks but I think add the option for wrapping each column could be useful too

mtias commented:
Unrelated to this, but I'm seeing child blocks in the quick block inserter outside the parent.

![image](https://user-images.githubusercontent.com/548849/41716439-b83573ac-7556-11e8-80f5-8b13a32f0bcb.png)

When inserting an item within a column, but not focusing the column block first, the block is appended at the end of the post:

![image](https://user-images.githubusercontent.com/548849/41716561-202c9710-7557-11e8-8cd6-9402f041bbb3.png)

mtias commented:
We should make each column block have the visual height of the container columns block.

![image](https://user-images.githubusercontent.com/548849/41716733-b2ad9260-7557-11e8-80f5-a6b7b2152ef8.png)

aduth committed:
Framework: Populate inner blocks template by insert if no blocks

aduth commented:
Pushed a few more updates.

- Fixed an issue where the initial Column blocks were not added when creating a new Columns (6dfe2af)
- Updated end-to-end tests per revised behavior (42e9201)
- Stretch Column to occupy full height of Columns parent (2050aa0)
   - Not entirely happy with the CSS here, but struggled for a while with various other approaches. Suggestions welcome.

SuperGeniusZeb commented:
@aduth I noticed that the Column block is not always the height of its parent, depending on what block you have selected:
![image](https://user-images.githubusercontent.com/19592990/41915470-6ba7d610-791b-11e8-99cf-e30f79f9ced8.png)
![image](https://user-images.githubusercontent.com/19592990/41915599-b78cdc6a-791b-11e8-9822-98413a30b1d5.png)

Is there any reason why you cannot just use some flex property like `align-items: stretch` to make the Column blocks fill the height of the Columns block?

aduth committed:
Testing: Update nested template E2E test

aduth committed:
Block API: Add support for valid block eligible deprecation

aduth committed:
Blocks: Add deprecated migration for Columns

aduth commented:
@SuperGeniusZeb The styling challenge is that the border is applied to `.editor-block-list__block-edit`, which is only one of many children within the column itself (along with toolbar, movers, sibling inserter, etc). I wasn't able to reproduce your original issue, though pushed a more Flexbox-y implementation in 003fab1 that may help your issue.

Another significant challenge I encountered is that with these changes, all existing Columns blocks would become invalid; worse, their content would disappear. This is because nested blocks were exempted from validation due to technical decisions around its unawareness of its own inner HTML (see https://github.com/WordPress/gutenberg/pull/3745#issuecomment-361746362, https://github.com/WordPress/gutenberg/pull/3745#discussion_r165011834). Our current deprecation behavior only takes effect for blocks which are strictly invalid. As a workaround, I added a new `isEligible` property function to a deprecation to allow a block to opt-in to a deprecated migration under specific circumstances even if the block is considered valid. With this, I was able to add a deprecated migration for old Columns content to the new format.

(cc @youknowriad)

GlennMartin1 commented:
I'm not a coder, but for what it's worth:

It's arguable if you should feel compelled to make columns backward-compatible, considering that it's been clearly designated as "beta." 

Columns need a lot of development. Don't compromise development by attempting backward-compatibility.

youknowriad reviewed (changes requested):
Code wise, this looks good. A noticed that it's possible to drag and drop a column block inside another column block. I guess this should be forbidden?

aduth commented:
@youknowriad For intents and purposes of this pull request, I'm relying on `<InnerBlocks allowedBlocks={ /* ... */ } />` and the block `parent` APIs to enforce the restriction. If there are failings there, I might be inclined to address separately, though I'll certainly look into what might be the cause of this. It may be something with drag-and-drop specifically, though I seem to also recall @westonruter mentioning similar issues with blocks being allowed to insert _themselves_ as a child of its own type, regardless of the `allowedBlocks` configuration.

westonruter commented:
> I seem to also recall @westonruter mentioning similar issues with blocks being allowed to insert themselves as a child of its own type, regardless of the allowedBlocks configuration.

That's right. I tested a configuration where I had nested blocks with `parent` restrictions like A > B > C. At the A level I should only have been able to inserted B blocks, but I recall being able to insert A blocks as well. Come to think of it, maybe it's because A doesn't have a `parent` bottom-up restriction and I should be using `allowedBlocks` top-down restriction instead? Or maybe that is only for the inserter and allowed as a block-level configuration.

The PR I was testing with: https://github.com/Automattic/amp-wp/pull/1215/files

mtias commented:
@youknowriad @aduth related? https://github.com/WordPress/gutenberg/pull/7212

youknowriad commented:
@mtias It looks like it, I'm approving as this was the main blocker for me.

SuperGeniusZeb commented:
@aduth The column height issue is still occurring for me. When I have selected a block in one column, the other column is the minimum height it can be (the height of its children plus some padding). If I select a block in the other column, it becomes the correct height. Tested on Firefox Beta.

jorgefilipecosta commented on core-blocks/columns/index.js:
```diff
@@ -86,7 +142,10 @@ export const settings = {
-					<InnerBlocks layouts={ getColumnLayouts( columns ) } />
+					<InnerBlocks
+						template={ getColumnsTemplate( columns ) }
+						templateLock
```
We are setting templateLock as boolean while for the templa_lock we use for CPT's we use a string that allows none, insert or all. A boolean does not allow us to differentiate between a lock all and a lock insert and I feel we should be consistent and also offer both options for innerBlock locking.
Another thing I notice is that we still show the remove button on columns even thougth it does not work (because of the lock).

Should we try to merge PR https://github.com/WordPress/gutenberg/pull/6991 before given that it offers the lock for all, insert and none and also addresses some UI problems?

aduth commented:
Rebased to resolve conflicts.

@SuperGeniusZeb Can you give it another try after da1b626 ? I'm still noticing one oddity in Firefox, which is a small gap that appears and disappears depending on if the Column block is selected and Contextual Toolbars are visible. This appears to be unrelated to these specific changes, more to do with conflicts with the toolbar margins in Firefox. I'll plan to open a separate issue.

@jorgefilipecosta I agree we should try to get #6991 merged ahead of this. Let me know when it's rebased and I'll give it another review.

SuperGeniusZeb commented:
@aduth It looks like the column height issue is fixed now. :+1:

jorgefilipecosta commented:
> @jorgefilipecosta I agree we should try to get #6991 merged ahead of this. Let me know when it's rebased and I'll give it another review.

Hi @aduth the PR was rebased, if anything is missing on the locking to handle this scenario feel free to comment correctly and I will improve the PR.

jasmussen commented:
Very nice work! 

I've been testing out this branch, also to see if any of the work I'm doing to refactor the block list stylesheet in #7365 is going to conflict. Looks like there are a few style issues, but nothing too bad, and I'll be able to catch that in a rebase if need be:

![screen shot 2018-07-02 at 10 46 48](https://user-images.githubusercontent.com/1204802/42154538-505d2f5c-7de6-11e8-86f1-cfcfa130cc66.png)

However the addition of a wrapper for each column introduces a new challenge. It's best explained in GIF form, I think. The following is the behavior in `master`:

![master](https://user-images.githubusercontent.com/1204802/42154581-729108aa-7de6-11e8-9c0b-abcd440f5598.gif)

Note how as we hover each block, the hierarchy is fairly simple, and note especially how when we select a paragraph, _the immediate parent_ has its outline shown. The purpose of this is to _visually indicate where you need to click, to select the parent_. Imagine a blockquote with a bunch of paragraphs — you can easily click a paragraph, and then the parent blockquote receives a highlight so you know where to click to select the whole thing.

Here's this branch:

![wrapper](https://user-images.githubusercontent.com/1204802/42154674-ac58296a-7de6-11e8-8a3d-ac9793c49a75.gif)

Note how the _column wrapper_ is now highlighted as the parent block. Which, of course, it is. You can even hover just the column itself, as if it's a block, but it doesn't appear that you can do anything with this "column" block. The breadcrumb shown on hover is also no longer accurate. More importantly, now when you click a paragraph block, the _column_ block receives an outline to indicate where to click to select it. 

This behavior isn't very useful to a user. Given that a user can't do anything on a "column" block, it feels like we should hide this block to the user completely. To put it differently, it feels like the behavior we have in `master` is the desirable one here. Can we combine the best of both worlds?

The pertinent bit to highlight the parent, I think, starts here: https://github.com/WordPress/gutenberg/blob/master/editor/components/block-list/block.js#L601 — @youknowriad helped with it.

chrisvanpatten commented:
> This behavior isn't very useful to a user. Given that a user can't do anything on a "column" block, it feels like we should hide this block to the user completely.

I agree completely. However to play devil's advocate, this could also have an impact depending on where future column settings are placed. For instance the ability to set the width of a column or its responsiveness settings.

I'm personally of the opinion that column-specific settings should actually be managed by the container Columns block: often column-specific settings aren't _truly_ column-specific, but dependent on the context of their siblings (e.g. making one column 75% width means the other column should be adjusted to 25%; the relationship is symbiotic).

However, if the longer term vision is that individual column child blocks should manage their own settings, then hiding a column container does pose a problem.

aduth commented:
You both covered it well @jasmussen and @chrisvanpatten . In its current form, I don't think expanding the column to the full height adds much value, because there are no settings for a column. And the obvious one (width) may be managed by the Columns wrapper. I could imagine column-specific settings at some point (e.g. background color?), though in its current form I don't think it should expand. Also worth noting this is a block-specific styling that doesn't apply generally to InnerBlocks, and can be tweaked for the Columns block specifically in the future.

I do think it's a general design issue we'll need to work through though where there are multiple levels of nesting, and each "level" may have its own settings and need to be reasonably selectable.

For now I'm inclined to drop the styling to expand the column to full height of its Columns wrapper. Does that sound reasonable?

aduth commented:
>This behavior isn't very useful to a user. Given that a user can't do anything on a "column" block, it feels like we should hide this block to the user completely. To put it differently, it feels like the behavior we have in master is the desirable one here. Can we combine the best of both worlds?

Re-reading this, I get the sense that we'd prefer for it to not be present at all. I'm not entirely sure how we do this in such a way that is generic for all use of InnerBlocks, where as noted above sometimes the intermediary blocks are useful. This seems more in line with what @chrisvanpatten is suggesting with the transparent block in his comment at https://github.com/WordPress/gutenberg/issues/6895#issuecomment-397604822 .

jorgefilipecosta commented:
The need for "transparent blocks" or a similar concept also appeared in the layout blocks PR https://github.com/WordPress/gutenberg/pull/7414#issuecomment-400998959.

aduth commented:
I toyed with it a bit in the branch, and I think the general idea could be fairly "simple" but not "easy" to implement, where [`BlockListBlock`'s rendering](https://github.com/WordPress/gutenberg/blob/35a839917b23ecb497435669577f14f7e49a8573/editor/components/block-list/block.js#L368-L578) has a shortcut condition to render only the contents of its `BlockCrashBoundary` when it's intended to be a passthrough. This is made complicated by various event handlings, `ref` assignment (e.g. `focusTabbable`), and styling considerations (e.g. expectation of a root `div` with `data-type`, `className`, etc). If it's something worth exploring, I'd suggest we do so as a separate task.

jasmussen commented:
Great discussion here. 

I think there are two aspects to it — one is ensuring that blocks that ship with core have a good user experience, not only because they ship with the software, but because they will serve as inspiration for other blocks to come. In this case, like I've noted, I think it would be nice to reduce the cognitive overhead by requiring a user to only configure the columns block itself, as well as any content inside each column.

I know this is way more difficult than it sounds, and I think it's fine to explore separately. This goal is implementation-agnostic and markup-agnostic, it's purely an ideal to consider.

But this discussion has also made clear that no matter how good a user experience we can build, there will be a need to handle situations with very deep nesting. This is a point that @SuperGeniusZeb has made a couple of times in the past, and it's increasingly clear that we need to accommodate less than ideal nesting situations. In that vein, we might want to look at #6459 again.

aduth commented:
Unless there's opposition, my plan is to merge this as-is about an hour from now. I've created #7694 to expand upon the idea of a passthrough block supports, have commented at https://github.com/WordPress/gutenberg/issues/6895#issuecomment-402199141 with respect to block nesting wrapper implementations, and have referenced to #6459 which I think may be worth reopening as it relates to navigating between deep nesting of blocks.

jorgefilipecosta reviewed (approved):
This is working correctly in my tests. The fact the columns block is not transparent adds some weight to the UI, but also adds the useful feature of being able to set a class in each column.
Awesome work bringing this changes in 👍

mtias commented:
Thanks @aduth for driving this update!

samikeijonen commented:
Wuhuu!!
